src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1224,6 +1224,27 @@ public Task<bool> HasSuccessfullyLoadedAsync(ProjectState project, CancellationT
             : project.HasAllInformation ? SpecializedTasks.True : SpecializedTasks.False;
     }
 
+    public SolutionCompilationState WithForceOnlyRequiredGenerators(ProjectId projectId)
+    {
+        if (!_projectIdToTrackerMap.TryGetValue(projectId, out var tracker))
+        {
+            // we don't have a tracker for this yet, create one with the correct policy
+            tracker = RoslynImmutableInterlocked.GetOrAdd(
```

Comment:
Looks like there's a race here -- what happens if there isn't a tracker, but something else creates it between the check and the GetOrAdd that doesn't meet our policy?